libgadget/uvbg.c

```diff
@@ -0,0 +1,783 @@
+    ptrdiff_t *slab_ix_start = UVBGgrids.slab_ix_start;
+
+    // create buffers on each rank which is as large as the largest LOGICAL allocation on any single rank
+    int buffer_size = 0;
+    for (int ii = 0; ii < nranks; ii++)
+        if (slab_nix[ii] > buffer_size)
+            buffer_size = (int)slab_nix[ii];
+
+    buffer_size *= UVBG_DIM * UVBG_DIM;
+    float *buffer_mass = fftwf_alloc_real((size_t)buffer_size);
+    float *buffer_stars_slab = fftwf_alloc_real((size_t)buffer_size);
+    float *buffer_sfr = fftwf_alloc_real((size_t)buffer_size);
+
+    // I am going to reuse the RegionInd member which is from petapm.  I
+    // *think* this is ok as we are doing this after the grav calculations.
+    // This is a potentially stupid way to do things anyway and will most
+    // definitely need to be changed! There is no way we should have to search
+    // all of the particles to find out what slab they sit on, and then loop
+    // through all particles again n_slab times!
+    double box_size = All.BoxSize;
+    for(int ii = 0; ii < PartManager->NumPart; ii++) {
+        if((!P[ii].IsGarbage) && (!P[ii].Swallowed) && (P[ii].Type < 5)) {
+            ptrdiff_t ix = pos_to_ngp(P[ii].Pos[0], box_size, UVBG_DIM);
+            P[ii].RegionInd = searchsorted(&ix, slab_ix_start, nranks, sizeof(ptrdiff_t), compare_ptrdiff, -1, -1);
+        } else {
+            P[ii].RegionInd = -1;
+        }
+    }
+
+
+    for (int i_r = 0; i_r < nranks; i_r++) {
+
+        const int ix_start = slab_ix_start[i_r];
+        const int nix = slab_nix[i_r];
+
+        // init the buffers
+        for (int ii = 0; ii < buffer_size; ii++) {
+            buffer_mass[ii] = (float)0.;
+            buffer_stars_slab[ii] = (float)0.;
+            buffer_sfr[ii] = (float)0.;
+        }
+
+        // copy the current stellar mass grid so we can work out the SFRs
+        // WATCH OUT: prev_stars is a union with J21 to save memory.  Until
+        // find_HII_bubbles is called again, J21 will be invalid!
+        const size_t grid_n_real = UVBG_DIM * UVBG_DIM * UVBG_DIM;
+        memcpy(UVBGgrids.prev_stars, UVBGgrids.stars, sizeof(float) * grid_n_real);
+
+        // fill the local buffer for this slab
+        // TODO(smutch): This should become CIC
+        unsigned int count_mass = 0;
+        for(int ii = 0; ii < PartManager->NumPart; ii++) {
+            if(P[ii].RegionInd == i_r) {
+                int ix = pos_to_ngp(P[ii].Pos[0], box_size, UVBG_DIM) - ix_start;
+                int iy = pos_to_ngp(P[ii].Pos[1], box_size, UVBG_DIM);
+                int iz = pos_to_ngp(P[ii].Pos[2], box_size, UVBG_DIM);
+
+                int ind = grid_index(ix, iy, iz, UVBG_DIM, INDEX_REAL);
+
+                buffer_mass[ind] += P[ii].Mass;
+                count_mass++;
+            }
+        }
+
+        message(0, "Added %d particles to mass grid.\n", count_mass);
+
+        // reduce on to the correct rank
+        if (this_rank == i_r) {
```

Comment:
I think you are trading space with time in a very unparallel way -- if the intent is to get a simple version working and if this turns out to be hard to get right, you can consider falling back to using each rank holds the full UV grid -- and use MPI_Allreduce to blindly add it up cross all ranks every UV step -- fitting in 2~3  400x400x400 grid shouldn't be too much pain.